### PR TITLE
Support reissuance transactions

### LINKED ISSUE #27: Support re-issuance transactions

### PR DESCRIPTION
**Pull request by louisinger**
**This PR fixes reissuances/issuances serialization and add the `addReissuances` function**

### AddReissuances

- implements the exact same behavior of [go-elements' AddReissuance](https://github.com/vulpemventures/go-elements/blob/71dd075222b5880722fcdb922fee3150cd5f383e/pset/updater.go#L641): add 1 input and 2 outputs.
- during blinding step, `tokenKey` becomes optional: we don't need it in reissuance cases.
- integration testing in `integration/issuances.test.ts`.

### Fix Issuance serialization

The PR fixes the [+1 error ](https://github.com/vulpemventures/liquidjs-lib/issues/27#issuecomment-934267291)described by @Janaka-Steph causing `hashForWitnessV0` to produces `out of bound` errors.

However, the error also comes from issuancesSize: we mixed `issuancesHash` and `issuance data` (not the same thing). So: 
- `issuanceSize` becomes `issuancesSize` (with an "s") and computes the size for issuances' hash (the +1 error fix is there).
- in `bufferSize` (size of hashForWitness) we use the input's issuance size (using a new `issuanceSize` function - without the "s") instead of the issuances' hash size. 

In conclusion: we write a 0x00 buffer ONLY in [issuances hashes](https://github.com/vulpemventures/liquidjs-lib/blob/24dd53300497836a9b7553a08b776b2167387197/ts_src/transaction.ts#L590). If the input do not contain issuance, we do not add [the 0x00 in the witnessHash](https://github.com/vulpemventures/liquidjs-lib/blob/24dd53300497836a9b7553a08b776b2167387197/ts_src/transaction.ts#L597), i.e `issuanceSize = 0` (I hope this is clear :rofl: ).

it closes #27 
it closes #28 

@altafan @Janaka-Steph @tiero please review

### CODE REVIEW

File: ts_src/issuance.ts [hunk 1]
```diff
@@ -206,3 +241,68 @@ export function validateAddIssuanceArgs(args: AddIssuanceArgs): void {
...
+  }
+
+  return false;
+}
+
+function isConfidentialWitnessUtxo(witnessUtxo: WitnessUtxo): boolean {
+  return (
+    witnessUtxo.rangeProof !== undefined &&
+    witnessUtxo.surjectionProof !== undefined
+  );
```
Comment (altafan) [inline ts_src/issuance.ts]: the utxo is confidential if its nonce is not [0x00]!

File: ts_src/issuance.ts [hunk 2]
```diff
@@ -20,6 +22,19 @@ export interface AddIssuanceArgs {
...
+  nonWitnessUtxo?: NonWitnessUtxo;
+  prevoutBlinder: Buffer;
+  entropy: Buffer;
+  assetAmount: number;
+  assetAddress: string;
+  tokenAmount: number;
+  tokenAddress: string;
+  precision: number;
+}
+
```
Comment (altafan) [inline ts_src/issuance.ts]: maybe we could move this and also AddIssuanceArgs to pbst?

Comment (tiero) [reply ts_src/issuance.ts]: we will need to refactor Psbt anyway, dont know if right to add stuff now

Comment (altafan) [reply ts_src/issuance.ts]: They belong to psbt indeed.. they are the input types for addIssuance and addReissuance methods of the Psbt object.. they shouldn't stay here, this Is a lower level (raw)... we don't want to have here possibly deprecated types if the Psbt module will be refactored. Better to move them like we did in go-elements.

**Comment by altafan** [comment]
@louisinger  I would like to see an intergration test where we make an unconfidential issuance, then, in another transaction, we make asset and token outputs confidential by sending to the same wallet but using confidential addresses, and finally, we try to reissue the asset with the conf prevout token.

**Comment by louisinger** [comment]
> @louisinger I would like to see an intergration test where we make an unconfidential issuance, then, in another transaction, we make asset and token outputs confidential by sending to the same wallet but using confidential addresses, and finally, we try to reissue the asset with the conf prevout token.

Done.

However, something disturbs me:
1. the token amount depends on `confidentialTokenFlag`. This flag will directly modify the token asset hash. 
2. from my understanding, this flag notifies if the token output is confidential or not (same behavior in go-elements).
3. but it does not make sense! because the assetHash is not related to blinding.
4. In such a case (where you issue an **unconfidential** token (so the flag is unconf) and then you want to transform it into a confidential output), you need to specify manually the confidential flag. I mean that we cannot deduce the flag according to the address (like we did in go-elements): we **must** specify the exact same flag used during the issuance step.

It is not really a "technical" problem: I've just added a `confidentialFlag` argument in `AddReissuanceArgs` to fix this. But we're getting to the point where I have to put "confidential flag = false" on a token that I'm going to blind :expressionless: 

So: do you know why the token asset hash depends on a "confidential flag"? Does it make sense for you? Maybe I missunderstand something @altafan

**Comment by altafan** [comment]
@louisinger I think the `confidentialFlag` field should be removed from both AddIssuanceArgs and AddReissuanceArgs.

The value of this bool can be inferred by the type of addresses defined in the mentioned objects.

For the issuance, Its value will be `true` if the addresses defined in the args are confidential. Hence, we must ensure that they are either confidential or unconf. No "mixed" cases allowed so to speak.

In the reissuance, instead, only the asset address will be defined, therefore there will be no token in the issuance of the input and we just ignore the confidential flag (it will be always `false`).

This is exactly what we do in [go-elements](https://github.com/vulpemventures/go-elements/blob/master/pset/updater.go#L548).

**Comment by louisinger** [comment]
> The value of this bool can be inferred by the type of addresses defined in the mentioned objects.

The problem with this is that you can provide a confidential address in `addIssuance` without blinding the transaction later. 

> No "mixed" cases allowed so to speak.

IMO liquidjs-lib (and go-elements) should cover all the cases allowed by elements node. And the "mixed" case is theoretically possible. You can issue an asset and blind only the token output for instance.

> In the reissuance, instead, only the asset address will be defined, therefore there will be no token in the issuance of the input and we just ignore the confidential flag (it will be always `false`).

Currently, go-elements do the opposite: always set the flag to `true` (confidential). See [here](https://github.com/vulpemventures/go-elements/blob/master/pset/updater.go#L676).

---
After some investigation on elements, I think I understand: the `confidentialFlag` is `true` **if and only if** the initial issuance is blinded. After all, it makes sense because the `Issuance` object set on inputs does not contain any asset hash (only the entropy). Then with entropy + flag, elements (and liquidjs and goelements) are able to re-compute the asset hashes (token and issued asset).

That's why we must specify the confidential flag in `AddReissuanceArgs` and `AddIssuanceArgs`. Setting up the flag to `true` in reissuance will fail in case of the initial issuance is not blinded (because elements has generated the token hash with `false` from entropy --> causing bad-txin-out error).

conclusion: `confidentialFlag` is not related to outputs but **only** to issuance.

**Comment by altafan** [comment]
Looking at the elements CLI, I noticed that the `blindrawtransaction` command has a new argument `blindissuances` that makes it possible to have unconf input issuances and conf asset/token outputs and viceversa. Previously, if the issuances were unconf, the outputs should have been unconf too and same for the "conf" case.

The design of go-elements comes from the older one of the elements CLI and needs to be updated.

So, you're totally right, we need the confidential flag to be explicitly set in the Add(Re)IssuanceArgs type somehow... 
Maybe we could call it `blindedIssuance` instead of `confidentialFlag`, what do you think @louisinger?

File: output.log [hunk 3]
```diff
@@ -0,0 +1,41 @@
+time="2021-09-09T12:29:25Z" level=fatal msg="dial tcp 172.23.0.2:5432: connect: connection refused"
```
Comment (tiero) [inline output.log]: remove this

**Review by tiero — COMMENTED** [review]
need to compile the js too?

File: ts_src/address.ts [hunk 4]
```diff
@@ -322,13 +322,15 @@ function isBlech32(address: string, network: Network): boolean {
 
 function decodeBlech32(address: string): AddressType {
   const blech32addr = fromBlech32(address);
-  switch (blech32addr.data.length) {
+  switch (blech32addr.data.length - 2) {
```
Comment (tiero) [inline ts_src/address.ts:325]: why we never did this before?

Comment (louisinger) [reply ts_src/address.ts:325]: this is a fix, because the `fromBlech32` function [adds two bytes](https://github.com/vulpemventures/liquidjs-lib/blob/f066e906b6942e3e5b980791a64dffa316b5f2b5/ts_src/address.ts#L80) to `blech32.data`. it causes `decodeBlech32` to fails every time because the data buffer has two additional bytes.
